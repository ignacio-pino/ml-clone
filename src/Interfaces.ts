export interface IProductListData {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;
  shipping: {
    free_shipping: boolean;
  };
}

export interface ISortDataList {
  sort: ISortData;
  available_sorts: ISortData[];
}

export interface ISortData {
  id: string;
  name: string;
}
export interface IProductCard {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;

  free_shipping: boolean;
}

export interface IFilterDataList {
  filters: IFilterList[];
  availableFilters: IFilterList[];
}
export interface IFilterList {
  id: string;
  name: string;
  type: string;
  values: IFilterListValues[];
}

export interface IFilterListValues {
  id: string;
  name: string;
  results?: number;
  path_from_root?: IPathFromRoot[];
}

export interface ICategoryData {
  query?: string;
  categoryInformation?: ICategoryInformation;
  totalResults: number;
  childrenCategories?: IPathFromRoot[];
}

export interface ICategoryInformation {
  id: string;
  name: string;
  path_from_root?: IPathFromRoot[];
}

export interface IPathFromRoot {
  id: string;
  name: string;
}

export interface IChildrenCategoryData {
  id: string;
  name: string;
  total_items_in_this_category: number;
}

export interface IHeaderCategoryData {
  categoryId: string;
  categoryName: string;
}
