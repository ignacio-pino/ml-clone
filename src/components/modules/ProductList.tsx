import ProductCard from "../common/ProductCard";
import { IProductListData } from "../../Interfaces";
import { Box } from "@chakra-ui/react";

export default function ProductList({
  productDataList,
}: {
  productDataList: IProductListData[];
}) {
  return (
    <Box px={"5%"}>
      {productDataList?.map(
        ({
          id,
          title,
          original_price,
          price,
          shipping,
          thumbnail,
          currency_id,
        }) => (
          <ProductCard
            id={id}
            title={title}
            original_price={original_price?.toString()}
            price={price.toString()}
            free_shipping={shipping.free_shipping}
            thumbnail={thumbnail}
            currency_id={currency_id}
          />
        )
      )}
    </Box>
  );
}
