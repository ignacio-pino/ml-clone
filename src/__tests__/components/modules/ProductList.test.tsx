import { render, screen } from "@testing-library/react";
import ProductList from "../../../components/modules/ProductList";
import { MemoryRouter } from "react-router-dom";
import { IProductListData } from "../../../Interfaces";

describe("ProductList Component", () => {
  const mockProductDataList = [
    {
      id: "1",
      title: "Test Product 1",
      original_price: "100",
      price: "90",
      shipping: { free_shipping: true },
      thumbnail: "test_image_url_1",
      currency_id: "USD",
    },
    {
      id: "2",
      title: "Test Product 2",
      original_price: "",
      price: "80",
      shipping: { free_shipping: false },
      thumbnail: "test_image_url_2",
      currency_id: "UYU",
    },
  ];

  it("renders the ProductCard for each product", () => {
    render(
      <MemoryRouter>
        <ProductList
          productDataList={mockProductDataList as IProductListData[]}
        />
      </MemoryRouter>
    );

    mockProductDataList.forEach((product) => {
      const titleElement = screen.getByText(product.title);
      expect(titleElement).toBeInTheDocument();
    });
  });

  it("passes the correct props to ProductCard", async () => {
    render(
      <MemoryRouter>
        <ProductList
          productDataList={mockProductDataList as IProductListData[]}
        />
      </MemoryRouter>
    );

    mockProductDataList.forEach(async (product) => {
      const card = await screen.findByTestId("productCard");
      expect(card).toHaveAttribute("id", product.id);
      expect(card).toHaveAttribute("title", product.title);
      expect(card).toHaveAttribute("price", product.price.toString());
      if (product.original_price) {
        expect(card).toHaveAttribute(
          "original_price",
          product.original_price.toString()
        );
      }
      expect(card).toHaveAttribute("thumbnail", product.thumbnail);
      expect(card).toHaveAttribute("currency_id", product.currency_id);
    });
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ProductList
          productDataList={mockProductDataList as IProductListData[]}
        />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
