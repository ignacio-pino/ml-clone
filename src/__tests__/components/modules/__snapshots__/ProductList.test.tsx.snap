// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductList Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1ax7sdm"
  >
    <div
      class="chakra-card css-1q6bp6q"
      data-testid="productCard"
      id="1"
    >
      <img
        alt="Test Product 1"
        class="chakra-image css-1bywttm"
        src="test_image_url_1"
      />
      <div
        class="chakra-stack css-1qstkuz"
      >
        <div
          class="chakra-card__body css-1knigk4"
        >
          <button
            class="chakra-button css-12b0kss"
            type="button"
          >
            <span
              class="css-1baulvz"
              tabindex="0"
            >
              Test Product 1
            </span>
          </button>
          <del
            class="chakra-stack css-h9jpb5"
          >
            <p
              class="chakra-text css-1nggxx3"
            >
              U$S 100
            </p>
            <p
              class="chakra-text css-qc7exc"
            />
          </del>
          <div
            class="chakra-stack css-usuplf"
          >
            <div
              class="chakra-stack css-u2zw5"
            >
              <b
                class="chakra-text css-4ww9f7"
              >
                U$S 90
              </b>
              <b
                class="chakra-text css-to2aqv"
              />
            </div>
            <p
              class="chakra-text css-19ff2qr"
            >
              10% OFF
            </p>
          </div>
          <mark
            class="chakra-text css-1d599vv"
          >
            Env√≠o gratis
          </mark>
        </div>
      </div>
    </div>
    <div
      class="chakra-card css-1q6bp6q"
      data-testid="productCard"
      id="2"
    >
      <img
        alt="Test Product 2"
        class="chakra-image css-1bywttm"
        src="test_image_url_2"
      />
      <div
        class="chakra-stack css-1qstkuz"
      >
        <div
          class="chakra-card__body css-1knigk4"
        >
          <button
            class="chakra-button css-12b0kss"
            type="button"
          >
            <span
              class="css-1baulvz"
              tabindex="0"
            >
              Test Product 2
            </span>
          </button>
          <div
            class="chakra-stack css-1hdaorj"
          >
            <div
              class="chakra-stack css-u2zw5"
            >
              <b
                class="chakra-text css-4ww9f7"
              >
                $ 80
              </b>
              <b
                class="chakra-text css-to2aqv"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
